test/bst/core_test.clj

```diff
@@ -2,250 +2,291 @@
+                                      :left {:data 4
+                                             :left {:data 3 :left nil :right nil :height 0}
+                                             :right nil
+                                             :height 1}
+                                      :height 2}))))
     (testing "with one node"
-      (is (= false (bst/is-left-case? {:root 9 :left nil :right nil}))))
+      (is (= false (bst/is-left-case? {:data 9 :left nil :right nil :height 0}))))
     (testing "with an empty tree"
       (is (= false (bst/is-left-case? {}))))))
 
-(deftest is-left-right-case-test
+(deftest is-left-right-case?-test
   (testing "Does right subtree of left child violates the BST property"
     (testing "with a non-empty tree"
-      (is (= true (bst/is-left-right-case? {:root 7
+      (is (= true (bst/is-left-right-case? {:data 7
                                             :right nil
-                                            :left {:root 5
+                                            :left {:data 5
                                                    :left nil
-                                                   :right {:root 6 :left nil :right nil}}}))))
+                                                   :right {:data 6 :left nil :right nil :height 0}
+                                                   :height 1}
+                                            :height 2}))))
     (testing "with one node"
-      (is (= false (bst/is-left-right-case? {:root 9 :left nil :right nil}))))
+      (is (= false (bst/is-left-right-case? {:data 9 :left nil :right nil :height 0}))))
     (testing "with an empty tree"
       (is (= false (bst/is-left-right-case? {}))))))
 
-(deftest is-right-left-case-test
+(deftest is-right-left-case?-test
   (testing "Does right subtree of left child violates the BST property"
     (testing "with a non-empty tree"
-      (is (= true (bst/is-right-left-case? {:root 2
+      (is (= true (bst/is-right-left-case? {:data 2
                                             :left nil
-                                            :right {:root 5
-                                                    :left {:root 4 :left nil :right nil}
-                                                    :right nil}}))))
+                                            :right {:data 5
+                                                    :left {:data 4 :left nil :right nil :height 0}
+                                                    :right nil
+                                                    :height 1}
+                                            :height 2}))))
     (testing "with one node"
-      (is (= false (bst/is-right-left-case? {:root 9 :left nil :right nil}))))
+      (is (= false (bst/is-right-left-case? {:data 9 :left nil :right nil :height 0}))))
     (testing "with an empty tree"
       (is (= false (bst/is-right-left-case? {}))))))
 
-(deftest is-right-right-case-test
+(deftest is-right-right-case?-test
```

Comment:
You should have a discussion with @ssrihari and @kirang89 about whether these tests are worth writing.

The public API of your BST should be something like this:

* Create a tree from a list of data.
* Insert into the tree.
* Remove a node from the tree.
* Check whether some data is present in the tree.
* ...and maybe some other miscellaneous operations like `height`.

Other functions, such as `is-left-left-case?` and `rotate-left` are implementation details. Testing them directly can be useful if they have substantial functionality, but it can also make your code brittle. Changing the implementation of your BST API will now probably mean changing these tests. 

Instead, you could test the public API instead, and also check that the tree is balanced after every insert / remove operation for a variety of different tree configurations. This is an indirect way of testing the same functionality in a less brittle fashion. 

The only functions that should need testing are create, insert, remove, `has?` and maybe `min-node` and `height`.

Reply:
This was the first project for which @PriyangaPKini wrote tests @jysandy. I specifically asked Priyanga to write tests for every function just to get the hang of it.

@PriyangaPKini Sandy's note is important, let's take some time to understand it, and apply the learning on the next project.

Reply:
Okay